src/key.js

```diff
@@ -968,13 +973,25 @@ export function generate(options) {
   function generateSecretKey() {
     secretKeyPacket = new packet.SecretKey();
     secretKeyPacket.algorithm = enums.read(enums.publicKey, options.keyType);
-    return secretKeyPacket.generate(options.numBits);
+    var material;
+    if (typeof(options.material) !== "undefined") {
+      material = options.material.key;
+    }
+    return secretKeyPacket.generate(options.numBits, options.curve, material);
   }
 
   function generateSecretSubkey() {
     secretSubkeyPacket = new packet.SecretSubkey();
-    secretSubkeyPacket.algorithm = enums.read(enums.publicKey, options.keyType);
-    return secretSubkeyPacket.generate(options.numBits);
+    var subkeyType = options.keyType;
+    if (subkeyType === enums.publicKey.ecdsa) {
+      subkeyType = enums.publicKey.ecdh;
+    }
+    var material;
+    if (typeof(options.material) !== "undefined") {
+      material = options.material.subkey;
+    }
+    secretSubkeyPacket.algorithm = enums.read(enums.publicKey, subkeyType);
+    return secretSubkeyPacket.generate(options.numBits, options.curve, material);
   }
 
   function wrapKeyObject() {
```

Comment:
Here wrapKeyObject has to be modified to remove hash/encryption algorithms that should not be used with a given curve. For example a p521 key should not be used with sha1, but with sha256. Also prefer_hash_algorithm need to reflect that also.